### PR TITLE
use CONNECT method for proxied HTTPS requests

### PR DESCRIPTION
**Pull request by jescholl**
HTTPS requests sent to proxies must use the `CONNECT` method or else they return a 501 "Unsupported Request Method and Protocol".

This affects Bluecoat and Squid for sure, probably others, see #154 for more background.

### CODE REVIEW

**Comment by shirkdog** [comment]
I would like this to be an option for the users to turn on and off, keeping the default method set to GET, and if you have proxy issues, you use the CONNECT method.

**Comment by jescholl** [comment]
My proposed change leaves the request method set to `GET` for all requests except those made over https _and_ through a proxy.


When a proxy receives a `GET` request it makes that same request on behalf of the client (with a few header modifications), and when a proxy receives a `CONNECT` request, it just opens a TCP socket between the client and the requested server and lets the two of them talk directly.

When a client sends a `GET` request for an HTTPS uri to a proxy, the proxy can do one of the following

1. attempt to process it as a normal `GET` request

    This will fail because the client is expecting to negotiate TLS with the proxy, which it can't do because the proxy doesn't have the private key for the requested site.

2. recognize that it's HTTPS and rewrite it to a `CONNECT` method

    This will probably fail because there is nothing in the RFC saying this is allowed so most clients won't handle it gracefully.

    Though, it might not even be possible to handle this sort of thing gracefully since after sending the `GET` request to the proxy the client would then need to detect that it had been shifted to `CONNECT` session (which it should have just used in the first place), then know to negotiate TLS after the `GET`, then send another `GET` to the destination server since the first one was to the proxy and cannot be injected into the TLS session.

3. reject the request with HTTP 501 "Unsupported Method"

    This is what Squid, and Bluecoat do.

#### TLDR; 

For a proxy to properly process an HTTPS request, it must be initiated with a `CONNECT` method.  Any proxy server that doesn't flat out reject a `GET` request to an HTTPS uri is probably trying to do something malicious.  

-----------------------------------------
For additional reading, Squid's wiki has a nice [write up](http://wiki.squid-cache.org/Features/HTTPS) about proxying over HTTPS. And if you want to go a little more in-depth, [RFC 2817](https://www.ietf.org/rfc/rfc2817.txt) describes how TLS should be used over HTTP. Sections 5.2 and 5.3, describe tunneling secured communications through a proxy with the `CONNECT` method.
